### PR TITLE
Stop using CFormatter in non-C backends

### PR DESCRIPTION
**Pull request by Manishearth**
CFormatter is mostly only used to get abi names. These are unaffected by backend attributes or other backend-specific things, and should not be obtained via a backend-specific formatter, they should be obtained via the HIR directly.

This should make changes like #546 less entangled, too. Unfortunately this conflicts with that PR, but it's not too big.

Now the only code that should need to invoke CFormatter is code that actually needs C-like code (e.g. the C++ backend)

cc @robertbastian

### CODE REVIEW

File: core/src/ast/types.rs [hunk 1]
```diff
@@ -53,11 +53,11 @@ impl CustomType {
         }
     }
 
-    /// The name of the destructor in C
-    pub fn dtor_name(&self) -> String {
-        let name = format!("{}_destroy", self.name());
-        self.attrs().abi_rename.apply(name.into()).into()
-    }
+    // /// The name of the destructor in C
```
Comment (robertbastian) [inline core/src/ast/types.rs:56]: .

File: feature_tests/cpp/include/ContiguousEnum.h [hunk 2]
```diff
@@ -24,7 +24,6 @@ namespace capi {
 extern "C" {
 #endif
 
-void ContiguousEnum_destroy(ContiguousEnum* self);
```
Comment (sffc) [inline feature_tests/cpp/include/ContiguousEnum.h:27]: Question: why is there a change in `Cpp` output?

**Review by sffc — APPROVED** [review]
Praise: I like this change. Makes more clear what things backends should be depending on.

File: feature_tests/cpp/include/ContiguousEnum.h (hunk 2 above)
Comment (Manishearth) [reply feature_tests/cpp/include/ContiguousEnum.h:27]: We removed dtors from (non opaque) structs and enums, they're not useful.

**Pull request merged**